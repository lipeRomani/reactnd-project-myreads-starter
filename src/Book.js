import React, {Component} from 'react';
import PropTypes from 'prop-types';

class Book extends Component {
    
    handleChange = e => {
        const shelfId = e.target.value;
        this.props.changeShelf(this.props.book, shelfId);
    }

    render() {
        return (
            <div className="book">
                <div className="book-top">
                    <div
                        className="book-cover"
                        style={{
                        width: 128,
                        height: 193,
                        backgroundImage: `url('${this.props.book.imageLinks ? this.props.book.imageLinks.smallThumbnail : ""}')`
                    }}></div>
                    <div className="book-shelf-changer" style={(this.props.book.shelf && this.props.book.shelf !== 'none') ? {backgroundColor : 'orange'} : {}}>
                        <select value={this.props.book.shelf ? this.props.book.shelf : 'none' } onChange={this.handleChange}>
                            <option value="none" disabled>Move to...</option>
                            <option value="currentlyReading">Currently Reading</option>
                            <option value="wantToRead">Want to Read</option>
                            <option value="read">Read</option>
                            <option value="none">None</option>
                        </select>
                    </div>
                </div>
                <div className="book-title">{this.props.book.title}</div>
                <div className="book-authors">{this.props.book.authors ? this.props.book.authors.join(", ") : ""}</div>
            </div>
        );
    }
}

Book.propTypes = {
    book : PropTypes.object.isRequired,
    changeShelf : PropTypes.func.isRequired
}

export default Book;
