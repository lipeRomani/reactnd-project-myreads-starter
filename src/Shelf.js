import React, {Component} from 'react';
import Book from './Book';
import PropTypes from 'prop-types';

class Shelf extends Component {

    hasBooks() {
        for (const book of  this.props.books) {
            if (book.shelf === this.props.shelfId) {
                return true
            }
        }
        return false;
    }

    countBooks() {
        let count = 0;
        for (const book of  this.props.books) {
            if (book.shelf === this.props.shelfId) {
                count++;
            }
        }
        return count;
    }

    render() {
        return (
            <div className="bookshelf">
                <h2 className="bookshelf-title">{this.props.title} ({this.countBooks()})</h2>
                <div className="bookshelf-books">
                    {!this.hasBooks() && (
                        <p className="list-empty-info" >No books in this shelf.</p>
                    )}
                    <ol className="books-grid">
                        {this.props.books.map((book) => {
                            return (
                                (book.shelf === this.props.shelfId && 
                                    <li key={book.id}>
                                        <Book book={book} changeShelf={this.props.changeShelf} />
                                    </li>
                                )
                            );
                        })}                    
                    </ol>
                </div>
            </div>
        );
    }
}

Shelf.propTypes = {
    title   : PropTypes.string.isRequired,
    shelfId : PropTypes.string.isRequired,
    books   : PropTypes.arrayOf(PropTypes.object).isRequired,
    changeShelf : PropTypes.func.isRequired
}

export default Shelf;