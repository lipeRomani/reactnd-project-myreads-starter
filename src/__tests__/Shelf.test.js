import React from 'react';
import {shallow, mount, render} from 'enzyme';
import Shelf from '../Shelf';
import Book from '../Book';

const emptyShelf = {
        title       : "Test Shelf",
        shelfId     : "wantToRead",
        books       : [],
        changeShelf : jest.fn()
    }

it ('Render without problem', () => {
    const wrapper = shallow(
        <Shelf 
            title={emptyShelf.title} 
            shelfId={emptyShelf.shelfId}
            books={emptyShelf.books}
            changeShelf={emptyShelf.changeShelf}
        />);

    expect(wrapper).toMatchSnapshot();
});

it ('Shelf without books show empty message', () => {
    const wrapper = mount(
          <Shelf 
                title={emptyShelf.title} 
                shelfId={emptyShelf.shelfId}
                books={emptyShelf.books}
                changeShelf={emptyShelf.changeShelf}
            />
        );
    expect(wrapper.find(".list-empty-info")).toHaveLength(1);
});

it ('Shelf with books show books', () => {
    const shelfId = 'wantToRead';
    const books = [
        {title: "teste", id : "2212", authors:["Name1", "Name2"], shelf : shelfId},
        {title: "teste2", id : "1111", authors:["Name1", "Name2"], shelf : shelfId}
    ]
    
    const wrapper = mount(
          <Shelf 
                title={emptyShelf.title} 
                shelfId={shelfId}
                books={books}
                changeShelf={emptyShelf.changeShelf}
            />
        );
    
    expect(wrapper.find("Book")).toHaveLength(2);

});

it ('Shelf with books show books', () => {
    const books = [
        {title: "teste", id : "2212", authors:["Name1", "Name2"], shelf : "wantToRead"},
        {title: "teste2", id : "1111", authors:["Name1", "Name2"], shelf : "wantToRead"}
    ]
    
    const wrapper = mount(
          <Shelf 
                title={emptyShelf.title} 
                shelfId={emptyShelf.shelfId}
                books={books}
                changeShelf={emptyShelf.changeShelf}
            />
        );
    
    expect(wrapper.find("Book")).toHaveLength(2);
});

it ('Shelf with books, count books', () => {
    const books = [
        {title: "teste", id : "2212", authors:["Name1", "Name2"], shelf : "wantToRead"},
        {title: "teste2", id : "1111", authors:["Name1", "Name2"], shelf : "wantToRead"}
    ]
    const title = "Want to read";

    const wrapper = mount(
          <Shelf 
                title={title} 
                shelfId={emptyShelf.shelfId}
                books={books}
                changeShelf={emptyShelf.changeShelf}
            />
        );
    
    expect(wrapper.find(".bookshelf-title").text()).toEqual(`${title} (${2})`);
});