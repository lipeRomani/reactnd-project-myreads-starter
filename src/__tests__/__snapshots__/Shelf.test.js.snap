exports[`test Render without problem 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="bookshelf">
    <h2
        className="bookshelf-title">
        Test Shelf
         (
        0
        )
    </h2>
    <div
        className="bookshelf-books">
        <If
            test={true}>
            <p
                className="list-empty-info">
                No books in this shelf
            </p>
        </If>
        <ol
            className="books-grid">
            
        </ol>
    </div>
</div>,
  "nodes": Array [
    <div
      className="bookshelf">
      <h2
            className="bookshelf-title">
            Test Shelf
             (
            0
            )
      </h2>
      <div
            className="bookshelf-books">
            <If
                  test={true}>
                  <p
                        className="list-empty-info">
                        No books in this shelf
                  </p>
            </If>
            <ol
                  className="books-grid">
                  
            </ol>
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Shelf
        books={Array []}
        changeShelf={[Function]}
        shelfId="wantToRead"
        title="Test Shelf" />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Shelf {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "books": Array [],
          "changeShelf": [Function],
          "shelfId": "wantToRead",
          "title": "Test Shelf",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="bookshelf">
          <h2
                    className="bookshelf-title">
                    Test Shelf
                     (
                    0
                    )
          </h2>
          <div
                    className="bookshelf-books">
                    <If
                              test={true}>
                              <p
                                        className="list-empty-info">
                                        No books in this shelf
                              </p>
                    </If>
                    <ol
                              className="books-grid">
                              
                    </ol>
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="bookshelf">
          <h2
                    className="bookshelf-title">
                    Test Shelf
                     (
                    0
                    )
          </h2>
          <div
                    className="bookshelf-books">
                    <If
                              test={true}>
                              <p
                                        className="list-empty-info">
                                        No books in this shelf
                              </p>
                    </If>
                    <ol
                              className="books-grid">
                              
                    </ol>
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Shelf
    books={Array []}
    changeShelf={[Function]}
    shelfId="wantToRead"
    title="Test Shelf" />,
}
`;
